crates/parser2/src/parser/expr_atom.rs

```diff
@@ -83,8 +84,22 @@ impl super::Parse for BlockExprScope {
                 .map(SyntaxKind::is_item_head)
                 .unwrap_or_default()
             {
-                parser.parse(ItemScope::default())?;
-                continue;
+                let is_item = parser.dry_run(|parser| {
+                    if parser.bump_if(SyntaxKind::UnsafeKw) {
+                        parser
+                            .current_kind()
+                            .map(SyntaxKind::is_item_head)
+                            .unwrap_or_default()
+                            && !parser.bump_if(SyntaxKind::UnsafeKw)
+                    } else {
+                        true
+                    }
+                });
+
+                if is_item {
+                    parser.parse(ItemScope::default())?;
+                    continue;
+                }
```

Comment:
`is_item_head` should be changed to not include `UnsafeKw`, because that's no longer a sure indicator of item-ness.

`dry_run` has some overhead; it would be better to only do a `dry_run` when the first token is known to be `unsafe`.

Reply:
I've addressed the second point, but left `is_item_head` and `is_expr_atom_head` as they are.